File: emulator/memory.py
#! /usr/bin/env python

DEFAULT_MEMORY_SIZE = 1024 * 1024
DEFAULT_BASE_ADDRESS = 0

def signByte(byte):
    if (byte & 0x80):
        return byte - 0x100
    else:
        return byte

def unsignByte(byte):
    if (byte < 0):
        return byte + 0x100
    else:
        return byte    

def signWord(word):
    if (word & 0x80000000):
        return word - 0x100000000
    else:
        return word

def unsignWord(word):
    if (word < 0):
        return word + 0x100000000
    else:
        return word

class Memory():
    """Representation of the virtual memory
       physical RAM and I/O (through bindings on callbacks)
       are stored here
    """
    def __init__(self, size=DEFAULT_MEMORY_SIZE, base_address=DEFAULT_BASE_ADDRESS):
        # Create the requested memory area
        self.a_memory = [0x00 for _ in xrange(size)]
        self.base_address = base_address
        self.upper_end = size + base_address

    def __getitem__(self, address):
    	return self.get_ubyte(address)

    def get_ubyte(self, address):
    	"""Get the byte at address as a unsigned number"""
        # Check if the address is part of the RAM
        if self.base_address <= address < self.upper_end:
            return unsignByte(self.a_memory[address - self.base_address])
        else:
            return 0x00

    def get_sbyte(self, address):
    	"""Get the byte at address as a signed number"""
        # Check if the address is part of the RAM
        if self.base_address <= address < self.upper_end:
            return signByte(self.a_memory[address - self.base_address])
        else:
            return 0x00

    def get_uword(self, address):
    	"""Get the 32bits word starting at address as a unsigned number"""
        # Make sure the address is not out of bounds
        if self.base_address <= address and address + 4 < self.upper_end:
            index = address - self.base_address
            word = self.a_memory[index]
            word |= self.a_memory[index + 1] << 8
            word |= self.a_memory[index + 2] << 16
            word |= self.a_memory[index + 3] << 24
            return unsignWord(word)
        else:
            return 0x00000000

    def get_sword(self, address):
    	"""Get the 32bits word starting at address as a signed number"""
        # Make sure the address is not out of bounds
        if self.base_address <= address and address + 4 < self.upper_end:
            index = address - self.base_address
            word = self.a_memory[index]
            word |= self.a_memory[index + 1] << 8
            word |= self.a_memory[index + 2] << 16
            word |= self.a_memory[index + 3] << 24
            return signWord(word)
        else:
            return 0x00000000

    def __setitem__(self, address, item):
    	self.set_byte(address, item)

    def set_byte(self, address, byte):
    	"""Set the byte at address in memory"""
        # Check if the address is part of the a_memory
        if self.base_address <= address < self.upper_end:
            self.a_memory[address - self.base_address] = byte & 0xFF

    def set_word(self, address, word):
    	"""Set the 32bits word starting at address"""
        if self.base_address <= address and address + 4 < self.upper_end:
            index = address - self.base_address
            self.a_memory[index] = word & 0xFF
            self.a_memory[index + 1] = (word>>8) & 0xFF
            self.a_memory[index + 2] = (word>>16) & 0xFF
            self.a_memory[index + 3] = (word>>24) & 0xFF

File: emulator/cpp/memory.cpp
#include <cstdlib>

#include "memory.hpp"

Memory::Memory(const unsigned size, const unsigned base_address)
	: memory_size(size), base_address(base_address)
{
	this->_a_memory = (char *)calloc(size, sizeof(char));
}

Memory::~Memory(void)
{
	free(this->_a_memory);
}

int Memory::__getitem__(const unsigned address)
{
	return get_ubyte(address);
}

void Memory::__setitem__(const unsigned address, const int item)
{
	set_byte(address, item);
}

unsigned int Memory::get_ubyte(const unsigned address)
{
	return this->get_sbyte(address) & 0xFF;
}

unsigned int Memory::get_uword(const unsigned address)
{
	return (unsigned int)this->get_sword(address);
}

int Memory::get_sbyte(const unsigned address)
{
	this->_mutex.lock();

	int item = 0;
	if (this->base_address <= address &&
		address < this->memory_size) {
		item = this->_a_memory[address - this->base_address];
	}

	this->_mutex.unlock();
	return item;
}

int Memory::get_sword(const unsigned address)
{
	this->_mutex.lock();

	int item = 0;
	if (this->base_address <= address &&
		address + sizeof(int) < this->memory_size) {
		item = *(int*)(this->_a_memory + address - this->base_address);
	}

	this->_mutex.unlock();
	return item;
}

void Memory::set_byte(const unsigned address, const int byte)
{
	this->_mutex.lock();

	if (this->base_address <= address &&
		address < this->memory_size) {
		this->_a_memory[address - this->base_address] = (char)byte;
	}

	this->_mutex.unlock();
}

void Memory::set_word(const unsigned address, const long word)
{
	this->_mutex.lock();

	if (this->base_address <= address &&
		address + sizeof(int) < this->memory_size) {
		*(int*)(this->_a_memory + address - this->base_address) = word;
	}

	this->_mutex.unlock();
}
